### PR TITLE
Fast synchronization

### PR DESCRIPTION
**Pull request by zvookin**
I have a few more changes I'm making, but this is fairly close. Want to start buildbot testing and get comments.

### CODE REVIEW

**Comment by zvookin** [comment]
Stuck between no C++11 in the runtime and COMDAT errors on Mac. We need to figure out how to get clang to compile without COMDATS -- that is to do whatever it does on Mac by default.

File: src/runtime/thread_pool_common.h [hunk 1]
```diff
@@ -57,7 +56,7 @@ struct work_queue_t {
     }
 
 };
-WEAK work_queue_t work_queue;
+ WEAK work_queue_t work_queue;
```
Comment (dsharletg) [inline src/runtime/thread_pool_common.h]: Remove added space

**Comment by abadams** [comment]
Do a merge with master and check "apps/linear_algebra/bin/halide_benchmarks sger 2080" on a z840

I'm seeing much lower performance on this branch than master on a large linux box.

**Comment by abadams** [comment]
For context: I'm trying to solve the mystery of why our sger slows down as you increase the thread count over 16 or so.

File: src/runtime/synchronization_common.h [hunk 2]
```diff
@@ -0,0 +1,934 @@
...
+}
+
+struct mutex_parking_control : parking_control {
+    uintptr_t *lock_state;
+
+    mutex_parking_control(uintptr_t *lock_state);
+};
+
+// Only used in parking -- lock_full.
+bool mutex_parking_control_validate(void *control, validate_action &action) {
```
Comment (dsharletg) [inline src/runtime/synchronization_common.h]: Should these be members of mutex_parking_control?

File: src/runtime/osx_yield.cpp [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#include "runtime_internal.h"
+
+extern "C" int swtch_pri(int);
+
+namespace Halide { namespace Runtime { namespace Internal {
+
+WEAK void halide_thread_yield() {
+    swtch_pri(0);
```
[the remark this answers is not in the document]
Comment (zvookin) [reply src/runtime/osx_yield.cpp]: Will look into it. I'm going from what some other implementations have chosen to do.

**Comment by zvookin** [comment]
@dsharletg look at the OS specific `thread_parker` classes. These use OS mechanisms to have the thread sleep for longer times. It starts with some number (currently 40) of spins and then if the lock is still held, it parks using the OS mechanism. This is also used for condition variable waits. In the nothing running case, all thread pool threads are in the OS' condition variable wait routine as they were before this change. (We can possibly move to using other OS primitives. I haven't totally convinced myself we can't just use a semaphore actually. This would buy us some simplicity and portability.)

There is some question as to whether this should be using the PAUSE instruction on x86 or similar on processors that support such. Perhaps as a first spin phase before calling the OS yield routine. (Also, adding support for hardware lock elision should be doable.)

File: src/runtime/synchronization_common.h (hunk 2 above)
Comment (zvookin) [reply src/runtime/synchronization_common.h]: I guess I could make them static members. Originally they were virtual members, but we can't use vtables in the runtime. The parking control dispatch is the only part of this I'm not happy with re: ease of understanding and code maintainability.

**Comment by zvookin** [comment]
Here are performance numbers across the full range of thread counts on an HP Z840: https://gist.github.com/zvookin/5eeb1dcc8bfbd5c0cf7dcf9bd24bf24f . (Two runs with each of master and fast_synchronization code.)

A few things show up here:

1) Using all hardware threads is a bad idea for this benchmark on this machine. My guess is faster packages are figuring out a way to automatically reduce the thread count or are manually tuned to use fewer threads. Alternatively they may be using the core count not the hardware thread count. Or possibly they have an algorithm which has different thread behavior.

2) The fast_synchronization branch has significantly higher peak performance. say 32.5 MFLOPS vs. 23.5 MFLOPS or so. These occur at about the same thread count. (Given the size of the effect here, I'd expect to get at least one more useful thread of work with faster synchronization, but I expect the story is dominated by memory bandwidth when the compute is actually happening, so maybe not.)

3) In the tail off, fast_synchronization fails more badly. I will take a look at this, but it is not clear this is a performance issue in the synchronization code. It could actually be faster and thus returning threads to fight for scarce memory bandwidth more quickly reducing overall performance. (I.e. if the mutex parked threads for longer here, the code might run faster.) It is also possible use of the PAUSE instruction or tuning the spin count might change this result.

Bottom line is even with this data, I'm not convinced fast synchronization is worse. Higher level, we probably need to figure out a way to throttle the number of threads on a per kernel basis or something.

**Comment by abadams** [comment]
I was seeing the opposite effect - 2x gflops on master relative to the branch at high thread counts. I'll check again. Would be nice if I just had the numbers flipped.

openblas seems to scale much better - it was hitting 60 gflops at 48 threads for me! Our peak was 27 gflops at ~16 threads. You can tune the number of threads with OPENBLAS_NUM_THREADS. So I don't think that openblas is just cleverly scaling down the number of threads - it's getting more peak flops out of the machine. Maybe it's numa-aware?

**Comment by zvookin** [comment]
Scratch my previous comment. fast_synchronization is actually faster at high thread count.

File: src/runtime/qurt_yield.cpp [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#include "runtime_internal.h"
+
+// TODO: what should we use here???
```
[the remark this answers is not in the document]
Comment (pranavb-ca) [reply src/runtime/qurt_yield.cpp]: I am asking the qurt guys. From the documentation, it wasn't obvious to me.

**Comment by zvookin** [comment]
I have verified that with trunk llvm, compiling for 32-bit ARM, this does generate calls to libatomic. I will figure out how to make that not happen.

**Comment by steven-johnson** [comment]
Failures now appear to be timeouts on some targets (including 64-bit) -- I'm gonna restart those tests on the (unlikely) event these are just overloaded buildbot blips, but more likely is there's an actual bug lurking somewhere still.

File: src/runtime/runtime_internal.h [hunk 5]
```diff
@@ -219,6 +217,12 @@ __attribute__((always_inline)) T reinterpret(const U &x) {
 
 extern WEAK __attribute__((always_inline)) int halide_malloc_alignment();
 
+struct halide_thread_parker {
```
Comment (steven-johnson) [inline src/runtime/runtime_internal.h]: Is this struct used anywhere?

**Comment by steven-johnson** [comment]
This is finally all clean. Are we ready to land it?

**Comment by zvookin** [comment]
I have one more minor change I want to make. It should be reviewed now though.

**Comment by zvookin** [comment]
This should be ready to roll now, pending review feedback of course.

File: src/runtime/posix_threads.cpp [hunk 6]
```diff
@@ -1,37 +1,57 @@
...
+extern int pthread_cond_signal(pthread_cond_t *cond);
+extern int pthread_cond_destroy(pthread_cond_t *cond);
+extern int pthread_mutex_init(pthread_mutex_t *mutex, const void *attr);
+extern int pthread_mutex_lock(pthread_mutex_t *mutex);
+extern int pthread_mutex_unlock(pthread_mutex_t *mutex);
+extern int pthread_mutex_destroy(pthread_mutex_t *mutex);
+
+// Indication on first machine is this is not any faster. (Probably not any slower either, but it's more complex so...)
+// Checking it in to try it on a couple other machines, but for code review purposes it will likely get deleted.
+#define TLS_CACHING_PARKER 0
```
Comment (abadams) [inline src/runtime/posix_threads.cpp]: We should make a decision one way or the other on this

**Comment by steven-johnson** [comment]
Looks like the Travis failure needs investigation:
```
cd bin/build/tmp ; /home/travis/build/halide/Halide/bin/host/generator_aot_variable_num_threads
/home/travis/build/halide/Halide/src/runtime/thread_pool_common.h:68 Assert failed: offset == sizeof(work_queue_t) && "Logic error in thread pool work queue initialization.\n"
bash: line 1: 29311 Aborted                 (core dumped) /home/travis/build/halide/Halide/bin/host/generator_aot_variable_num_threads
make: *** [generator_aot_variable_num_threads] Error 134
```

**Comment by abadams** [comment]
Windows performance tests:
https://gist.github.com/abadams/d77665c791678d97e07803aca91908cc

TL;DR: performance_inner_loop_parallel is way faster, everything else is in the noise.

There are surprisingly large differences between mingw and ubuntu-on-windows-10, with neither being clearly faster. I'm guessing those might be differences in the allocators?

**Comment by abadams** [comment]
Looks like we're just missing the win-32 bot

**Comment by steven-johnson** [comment]
It's all clean. Are we ready to merge?

**Comment by abadams** [comment]
Zalman mentioned maybe wanting to do a release first, but it's all good from my point of view.

**Comment by steven-johnson** [comment]
Pretty sure we can do a release from an arbitrary hash, so that's not a problem. On Gitter there was some suggestion we wait a week or two. I have no big preference other than trying to do them more regularly than we have been.

**Comment by abadams** [comment]
We have all this fabulous new machinery and we're just using it to reimplement condition variables and mutexes. Looking ahead should we be using the parking lot abstraction more directly to coordinate our worker threads?

**Comment by steven-johnson** [comment]
Baby steps, @abadams :-)
